[Explicit Module Builds] C++ Interoperability mode fixes

artemcm opened the pull request.

artemcm committed:
[Explicit Module Builds] Propagate the C++ Interop mode to interface sub-invocations and dependency scanner

hyp commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1801,6 +1801,24 @@ InterfaceSubContextDelegateImpl::InterfaceSubContextDelegateImpl(
+      compatVersion = "swift-5.9";
+    else if (langOpts.cxxInteropCompatVersion[0] == version::getUpcomingCxxInteropCompatVersion())
+      compatVersion = "upcoming-swift";
+    else // TODO: This may need to be updated once more versions are added
```
Can you leave a comment in `validateCxxInteropCompatibilityMode` so that next time we update it we don't forget this.

artemcm commented:
@swift-ci smoke test

artemcm commented:
@swift-ci smoke test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci smoke test

artemcm committed:
[Dependency Scanning] Add required additional C++ interop overlay module queries

artemcm commented:
@swift-ci test

artemcm committed:
Always add an implicit import of 'Cxx' module when C++ Interop is enabled.

We cannot always rely on being able to do so only as an overlay query upon loading 'requires cplusplus' modulemap modules. The 'requires' statement only applies to submodules, and we may not be able to query language feature modulemap attributes in dependency scanning context.

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

artemcm committed:
Allow users to opt-out of implicit `Cxx` import with `-disable-implicit-cxx-module-import`

artemcm commented:
@swift-ci test

artemcm commented:
@egorzhdan ping for review.

egorzhdan commented on include/swift/DependencyScan/SerializedModuleDependencyCacheFormat.h:
```diff
@@ -193,7 +195,8 @@ using ClangModuleDetailsLayout =
                    IdentifierIDField,         // clangIncludeTreeRoot
-                   IdentifierIDField          // moduleCacheKey
+                   IdentifierIDField,         // moduleCacheKey
+                   IsSystemField              // isSystem
```
Not my area of expertise 😀 but does this require bumping `MODULE_DEPENDENCY_CACHE_FORMAT_VERSION_MAJOR`?

artemcm replied:
Good catch. Fixed, thank you.

egorzhdan commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1969,6 +1969,31 @@ InterfaceSubContextDelegateImpl::InterfaceSubContextDelegateImpl(
+    std::string compatVersion;
+    if (langOpts.cxxInteropCompatVersion.empty())
+      compatVersion = "default";
+    else if (langOpts.cxxInteropCompatVersion[0] == 5)
```
Since we also have `swift-6` interop compat version now, should it also be handled here?

artemcm replied:
Done!

egorzhdan commented on test/Interop/Cxx/foreign-reference/witness-table.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-run-simple-swift(-I %S/Inputs/ -Xfrontend -enable-experimental-cxx-interop -Xfrontend -validate-tbd-against-ir=none -Xfrontend -disable-llvm-verify -Xfrontend -disable-availability-checking -g)
+// RUN: %target-run-simple-swift(-I %S/Inputs/ -Xfrontend -enable-experimental-cxx-interop -Xfrontend -validate-tbd-against-ir=none -Xfrontend -disable-llvm-verify -Xfrontend -disable-availability-checking -Xfrontend -disable-implicit-cxx-module-import -g)
```
I think we shouldn't need the extra flag here anymore after my recent changes to this test.

egorzhdan commented on test/Interop/Cxx/library-evolution/allow-objc-in-cxx-mode-in-evolving-libraries.swift:
```diff
@@ -1,6 +1,6 @@
 // RUN: %empty-directory(%t)
 // RUN: split-file %s %t
-// RUN: %target-swift-frontend %t/test.swift -I %t/Inputs -typecheck -enable-library-evolution -enable-experimental-cxx-interop -verify
+// RUN: %target-swift-frontend %t/test.swift -I %t/Inputs -typecheck -enable-library-evolution -enable-experimental-cxx-interop -disable-implicit-cxx-module-import -verify
```
Why is the `disable-implicit-cxx-module-import` flag required in this test?

artemcm replied:
```
<unknown>:0: error: unexpected warning produced: module 'Cxx' was not compiled with library evolution support; using it means binary compatibility for 'test' can't be guaranteed
```
Because we import `Cxx` module when enabling interop, enabling library evolution has to mean disabling this implicit import.

artemcm replied:
And because I don't know how to handle error messages without source-location with `-verify` I thought I'd just disable it.

artemcm replied:
I've refactored the decision to implicit import `Cxx` into a new `shouldImportCxx` method and made it default to `false` for when Library Evolution is enabled.

egorzhdan replied:
Ah, I see! Thanks for explaining

egorzhdan reviewed (commented):
Thanks @artemcm! I only have a few minor questions.

artemcm commented:
@swift-ci test

egorzhdan reviewed (approved):
LGTM!
